processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -225,13 +236,19 @@ protected AddToFactsResult addToFacts(
    * </ul>
    *
    * @param key                          TimeAndDims key
-   * @param maxBytesPerRowForAggregators max size per aggregator
-   *
+   * @param maxBytesPerRowForAggregators max size per row for aggregators
+   * @param actualRowSizeForAggregators  actual aggregator size for this row
    * @return estimated size of row
    */
-  private long estimateRowSizeInBytes(IncrementalIndexRow key, long maxBytesPerRowForAggregators)
+  private long estimateRowSizeInBytes(
+      IncrementalIndexRow key,
+      long maxBytesPerRowForAggregators,
+      long actualRowSizeForAggregators
+  )
   {
-    return ROUGH_OVERHEAD_PER_MAP_ENTRY + key.estimateBytesInMemory() + maxBytesPerRowForAggregators;
+    return ROUGH_OVERHEAD_PER_MAP_ENTRY
+           + key.estimateBytesInMemory()
+           + (useMaxMemoryEstimates ? maxBytesPerRowForAggregators : actualRowSizeForAggregators);
```

Comment:
In reading this code I found myself wondering why push down the choice of which value to use so low and add an extra parameter here instead of change the call site of `estimateRowSizeInBytes` to check and pass the "correct" value?

Reply:
Fixed.